### ISSUE TITLE
StackOverflowError on cyclic dependency with deepImmutablesDetection = true

### ISSUE THREAD

**Issue by scheleaap** [immutables/immutables#549]
If a cyclic dependency exists between immutable types annotated with `@Value.Style(deepImmutablesDetection = true)` a StackOverflowError occurs during compilation.

### Steps to reproduce

Class A:
```java
package test;

import org.immutables.value.Value;

@Value.Immutable
@Value.Style(deepImmutablesDetection = true)
public interface A {

    B getB();

}
```

Class B:
```java
package test;

import org.immutables.value.Value;

@Value.Immutable
@Value.Style(deepImmutablesDetection = true)
public interface B {

    A getA();

}
```

Result:
```
Caused by: java.lang.StackOverflowError
	at org.immutables.value.internal.$processor$.meta.$ImmutableProto$MetaAnnotated.<init>($ImmutableProto.java:1340)
	at org.immutables.value.internal.$processor$.meta.$ImmutableProto$MetaAnnotated.of($ImmutableProto.java:1728)
	at org.immutables.value.internal.$processor$.meta.$Proto$MetaAnnotated.from($Proto.java:205)
	at org.immutables.value.internal.$processor$.meta.$Proto$AbstractDeclaring.metaAnnotated($Proto.java:506)
	at org.immutables.value.internal.$processor$.meta.$ImmutableProto$DeclaringType.metaAnnotated($ImmutableProto.java:3386)
	at org.immutables.value.internal.$processor$.meta.$Proto$DeclaringType.isEnclosing($Proto.java:924)
	at org.immutables.value.internal.$processor$.meta.$ImmutableProto$DeclaringType.isEnclosing($ImmutableProto.java:3026)
	at org.immutables.value.internal.$processor$.meta.$Proto$DeclaringType.validate($Proto.java:1003)
	at org.immutables.value.internal.$processor$.meta.$ImmutableProto$DeclaringType.validate($ImmutableProto.java:3587)
	at org.immutables.value.internal.$processor$.meta.$ImmutableProto$DeclaringType.access$3300($ImmutableProto.java:2634)
	at org.immutables.value.internal.$processor$.meta.$ImmutableProto$DeclaringType$Builder.build($ImmutableProto.java:3681)
	at org.immutables.value.internal.$processor$.meta.$Round.declaringTypeFrom($Round.java:168)
	at org.immutables.value.internal.$processor$.meta.$Round$ProtoclassCollecter.collectIncludedAndDefinedBy($Round.java:282)
	at org.immutables.value.internal.$processor$.meta.$Round$ProtoclassCollecter.collect($Round.java:186)
	at org.immutables.value.internal.$processor$.meta.$Round$ProtoclassCollecter.collect($Round.java:176)
	at org.immutables.value.internal.$processor$.meta.$Round.protoclassesFrom($Round.java:104)
	at org.immutables.value.internal.$processor$.meta.$Proto$Environment.protoclassesFrom($Proto.java:379)
	at org.immutables.value.internal.$processor$.meta.$ValueAttribute.initAttributeValueType($ValueAttribute.java:1062)
	at org.immutables.value.internal.$processor$.meta.$ValueAttribute.initAndValidate($ValueAttribute.java:973)
	at org.immutables.value.internal.$processor$.meta.$AccessorAttributesCollector.collect($AccessorAttributesCollector.java:103)
	at org.immutables.value.internal.$processor$.meta.$ValueTypeComposer.compose($ValueTypeComposer.java:64)
	at org.immutables.value.internal.$processor$.meta.$Proto$Environment.composeValue($Proto.java:375)
	at org.immutables.value.internal.$processor$.meta.$ValueAttribute.initAttributeValueType($ValueAttribute.java:1066)
	at org.immutables.value.internal.$processor$.meta.$ValueAttribute.initAndValidate($ValueAttribute.java:973)
	at org.immutables.value.internal.$processor$.meta.$AccessorAttributesCollector.collect($AccessorAttributesCollector.java:103)
	at org.immutables.value.internal.$processor$.meta.$ValueTypeComposer.compose($ValueTypeComposer.java:64)
	at org.immutables.value.internal.$processor$.meta.$Proto$Environment.composeValue($Proto.java:375)
	at org.immutables.value.internal.$processor$.meta.$ValueAttribute.initAttributeValueType($ValueAttribute.java:1066)
	at org.immutables.value.internal.$processor$.meta.$ValueAttribute.initAndValidate($ValueAttribute.java:973)
... and so on
```

The problem applies to more complex dependency graphs as well (e.g. `A` -> `B` -> `C` -> `A`).

### Expected behavior

It would be great if Immutables would stop traversing the dependency graph if it encounters a node it has seen before already.

**Comment by elucash** [comment]
have to be fixed, thank you for reporting this!

**Comment by elucash** [comment]
released in 2.4.4. thank you!
